### PR TITLE
Don't recycle volumes that are still in use

### PR DESCRIPTION
**Pull request by mhutter**

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+.git
```
Comment (hansmi) [inline .dockerignore]: Doesn't `.dockerignore` support absolute paths, e.g. `/.git/`?

File: recycler.sh [hunk 2]
```diff
@@ -254,6 +254,21 @@ recreate_volume() {
...
 
+# Verify that a volume is not mounted anywhere. Returns the number of fuse
+# clients that are mounting the specified volume.
+#
+# $1 - list of gluster endpoints
+# $2 - name of the volume
+volume_unused() {
+  local rhost="${1%%,*}"  # extract the first entry
+  local vol_name="$2"
+  nfuse="$(gluster --remote-host="$rhost" volume status "$vol_name" client-list | \
```
Comment (hansmi) [inline recycler.sh]: Use `--mode=script` just in case.

File: recycler.sh [hunk 3]
```diff
@@ -254,6 +254,21 @@ recreate_volume() {
   return 0
 }
 
+# Verify that a volume is not mounted anywhere. Returns the number of fuse
+# clients that are mounting the specified volume.
+#
+# $1 - list of gluster endpoints
+# $2 - name of the volume
+volume_unused() {
```
Comment (hansmi) [inline recycler.sh]: Generally speaking "positive" names are easier to understand: `volume_in_use`.

File: recycler.sh [hunk 4]
```diff
@@ -254,6 +254,21 @@ recreate_volume() {
...
+  local rhost="${1%%,*}"  # extract the first entry
+  local vol_name="$2"
+  nfuse="$(gluster --remote-host="$rhost" volume status "$vol_name" client-list | \
+    awk '/^\w+\s+[[:digit:]]+$/{print $2}')"
```
Comment (hansmi) [inline recycler.sh]: Doesn't this also match the last line with the summary? How does this handle the self-heal daemon?

File: recycler.sh [hunk 5]
```diff
@@ -254,6 +254,21 @@ recreate_volume() {
...
+
+  # nfuse should be completely empty if there are no mounts
+  return "$((nfuse+0))"
```
Comment (hansmi) [inline recycler.sh]: Why `+0`? Also, the return code has a limited range and counts larger than 127 will not have the desired effect. Return `0` or `1` explicitly.

**Review by hansmi — CHANGES_REQUESTED** [review]

File: .dockerignore (hunk 1 above)
Comment (mhutter) [reply .dockerignore]: Indeed, adjusted

File: recycler.sh (hunk 4 above)
Comment (mhutter) [reply recycler.sh]: Good catch, copy/paste fail

File: recycler.sh [hunk 6]
```diff
@@ -254,6 +254,26 @@ recreate_volume() {
...
+volume_is_used() {
+  local rhost="${1%%,*}"  # extract the first entry
+  local vol_name="$2"
+  nfuse="$(
+    gluster --remote-host="$rhost" --mode=script \
+      volume status "$vol_name" client-list | \
+    awk '/^fuse\s+[[:digit:]]+$/{print $2}')"
+
+  # if not mounted, nfuse should be empty (which is `-eq` to 0)
+  if [ "$nfuse" -eq 0 ]; then
```
Comment (hansmi) [inline recycler.sh]: Use Bash syntax, like the rest of the code: `[[ … ]]`.

File: recycler.sh [hunk 7]
```diff
@@ -254,6 +254,26 @@ recreate_volume() {
...
+#
+# $1 - list of gluster endpoints
+# $2 - name of the volume
+volume_is_used() {
+  local rhost="${1%%,*}"  # extract the first entry
+  local vol_name="$2"
+  nfuse="$(
+    gluster --remote-host="$rhost" --mode=script \
+      volume status "$vol_name" client-list | \
+    awk '/^fuse\s+[[:digit:]]+$/{print $2}')"
```
Comment (hansmi) [inline recycler.sh]: This will ignore any non-FUSE clients. What's the rationale for not counting all clients except `glustershd`?

**Review by hansmi — CHANGES_REQUESTED** [review]

File: recycler.sh (hunk 7 above)
Comment (mhutter) [reply recycler.sh]: TODO: Check what the docs have to say about possible outputs; change code to count non-glustershd-clients

File: recycler.sh [hunk 8]
```diff
@@ -254,6 +254,26 @@ recreate_volume() {
...
+# Verifies that no fuse clients are connected to the volume given
+#
+# $1 - list of gluster endpoints
+# $2 - name of the volume
+volume_is_used() {
+  local rhost="${1%%,*}"  # extract the first entry
+  local vol_name="$2"
+  nfuse="$(
+    gluster --remote-host="$rhost" --mode=script \
+      volume status "$vol_name" client-list | \
```
Comment (mhutter) [inline recycler.sh]: Does not work with `glusterfs 3.12.15`:

    No brick client-list in volume gluster-pv66

Comment (mhutter) [reply recycler.sh]: We'll have to test what this means for 3.12 and lower clusters; depending on whether this is implemented in the client or the server we might be good...

File: recycler.sh (hunk 7 above)
Comment (mhutter) [reply recycler.sh]: Reimplemented `volume_is_used` to handle all types of clients (as long as they match `\w+`)

File: recycler.sh (hunk 8 above)
Comment (mhutter) [reply recycler.sh]: When used with pre-4.0 gluster cluster, the `client-list` command will return an empty list. This retains the current behavior of the recycler, but does not break it.

**Review by hansmi — APPROVED** [review]

**Pull request merged**